diff --git a/net/http.rb b/net/http.rb
index 88a174a248..4da8dfa894 100644
--- a/net/http.rb
+++ b/net/http.rb
@@ -589,7 +589,7 @@ def HTTP.start(address, *arg, &block) # :yield: +http+
       port, p_addr, p_port, p_user, p_pass = *arg
       p_addr = :ENV if arg.size < 2
       port = https_default_port if !port && opt && opt[:use_ssl]
-      http = new(address, port, p_addr, p_port, p_user, p_pass)
+      http = new(address, port, p_addr, p_port, p_user, p_pass, (ENV['NO_PROXY']||ENV['no_proxy']))
       http.ipaddr = opt[:ipaddr] if opt && opt[:ipaddr]
 
       if opt
@@ -642,7 +642,8 @@ def HTTP.new(address, port = nil, p_addr = :ENV, p_port = nil, p_user = nil, p_p
       elsif p_addr == :ENV then
         http.proxy_from_env = true
       else
-        if p_addr && p_no_proxy && !URI::Generic.use_proxy?(p_addr, p_addr, p_port, p_no_proxy)
+        # patched: https://github.com/ruby/net-http/pull/64
+        if p_addr && p_no_proxy && !URI::Generic.use_proxy?(address, address, p_port, p_no_proxy)
           p_addr = nil
           p_port = nil
         end
